File: cpp/samples/core/gmod.cpp
/**
 * @file gmod.cpp
 * @brief Demonstrates usage of vista-sdk Gmod API
 * @details This sample shows how to access the Generic Product Model (Gmod),
 *          navigate nodes, query metadata, and work with product types/selections
 */

#include <dnv/vista/sdk/core/VIS.h>

#include <iomanip>
#include <iostream>
#include <string>
#include <unordered_set>
#include <vector>

int main()
{
    using namespace dnv::vista::sdk;

    std::cout << "=== vista-sdk Gmod Sample ===\n\n";

    {
        std::cout << "1. Gmod: Accessing Gmod for a VIS version\n";
        std::cout << "--------------------------------------------\n";

        const auto& vis = VIS::instance();
        auto version = vis.latest();
        const auto& gmod = vis.gmod(version);

        std::cout << "Gmod for version   : " << VisVersions::toString(gmod.version()) << "\n";

        size_t nodeCount = 0;
        for ([[maybe_unused]] const auto& [code, node] : gmod)
        {
            nodeCount++;
        }

        std::cout << "Total nodes in Gmod: " << nodeCount << "\n";
        std::cout << "Root node code     : " << gmod.rootNode().code() << "\n";
        std::cout << "\n";
    }

    {
        std::cout << "2. GmodNode: Accessing nodes by code\n";
        std::cout << "---------------------------------------\n";

        const auto& vis = VIS::instance();
        const auto& gmod = vis.gmod(vis.latest());

        // Access node using operator[]
        const auto& node1 = gmod["411.1"];
        std::cout << "Node code    : " << node1.code() << "\n";
        std::cout << "Node name    : " << node1.metadata().name() << "\n";
        std::cout << "Node category: " << node1.metadata().category() << "\n";
        std::cout << "Node type    : " << node1.metadata().type() << "\n";

        // Safe access using node
        auto node2Opt = gmod.node("C101.31");
        if (node2Opt.has_value())
        {
            std::cout << "\nFound node: " << (*node2Opt)->code() << "\n";
            std::cout << "Name      : " << (*node2Opt)->metadata().name() << "\n";
        }

        // Try invalid code
        auto invalidOpt = gmod.node("INVALID");
        std::cout << "\nInvalid code lookup: " << std::boolalpha << invalidOpt.has_value() << "\n";

        std::cout << "\n";
    }

    {
        std::cout << "3. GmodNode: Navigating the tree hierarchy\n";
        std::cout << "---------------------------------------------\n";

        const auto& vis = VIS::instance();
        const auto& gmod = vis.gmod(vis.latest());

        const auto& node = gmod["411.1"];

        // Parents (can have multiple in Gmod graph)
        if (!node.parents().empty())
        {
            std::cout << "Parents of " << node.code() << " (" << node.parents().size() << "):\n";
            for (const auto* parent : node.parents())
            {
                std::cout << "  - " << parent->code() << "\n";
            }
        }

        // Children
        std::cout << "\nChildren of " << node.code() << " (" << node.children().size() << "):\n";
        size_t maxDisplay = 5;
        size_t count = 0;
        for (const auto* child : node.children())
        {
            if (count++ >= maxDisplay)
            {
                std::cout << "  ... and " << (node.children().size() - maxDisplay) << " more\n";
                break;
            }
            std::cout << "  - " << child->code() << " (" << child->metadata().name() << ")\n";
        }

        // Root check
        std::cout << "\nIs root  ? " << std::boolalpha << node.isRoot() << "\n";
        std::cout << "Root node: " << gmod.rootNode().code() << "\n";
        std::cout << "Is root  ? " << std::boolalpha << gmod.rootNode().isRoot() << "\n";

        std::cout << "\n";
    }

    {
        std::cout << "4. GmodNode: Node metadata\n";
        std::cout << "----------------------------\n";

        const auto& vis = VIS::instance();
        const auto& gmod = vis.gmod(vis.latest());

        const auto& node = gmod["S222.11"];
        const auto& metadata = node.metadata();

        std::cout << "Node                : " << node.code() << "\n";
        std::cout << "Category            : " << metadata.category() << "\n";
        std::cout << "Type                : " << metadata.type() << "\n";
        std::cout << "Name                : " << metadata.name() << "\n";

        if (metadata.definition().has_value())
        {
            std::cout << "Definition      : " << metadata.definition().value() << "\n";
        }

        if (metadata.installSubstructure().has_value())
        {
            std::cout << "Install substructure: " << metadata.installSubstructure().value() << "\n";
        }

        std::cout << "\n";
    }

    {
        std::cout << "5. GmodNode: Product types and selections\n";
        std::cout << "--------------------------------------------\n";

        const auto& vis = VIS::instance();
        const auto& gmod = vis.gmod(vis.latest());

        // Product type example
        const auto& node1 = gmod["411.3"];
        std::cout << "Node             : " << node1.code() << "\n";
        std::cout << "Has product type : " << std::boolalpha << node1.productType().has_value() << "\n";
        if (auto prodTypeOpt = node1.productType())
        {
            const GmodNode* prodType = prodTypeOpt.value();
            std::cout << "Product type code: " << prodType->code() << "\n";
            std::cout << "Product type name: " << prodType->metadata().name() << "\n";
        }

        // Product selection example
        const auto& node2 = gmod["411.2"];
        std::cout << "\nNode: " << node2.code() << "\n";
        std::cout << "Has product selection : " << std::boolalpha << node2.productSelection().has_value() << "\n";
        if (auto prodSelOpt = node2.productSelection())
        {
            const GmodNode* prodSel = prodSelOpt.value();
            std::cout << "Product selection code: " << prodSel->code() << "\n";
            std::cout << "Product selection name: " << prodSel->metadata().name() << "\n";
        }

        // Check if node is product selection
        auto csNode = gmod.node("CS1");
        if (csNode.has_value())
        {
            std::cout << "\nNode: " << (*csNode)->code() << "\n";
            std::cout << "Is product selection: " << std::boolalpha << (*csNode)->isProductSelection() << "\n";
        }

        std::cout << "\n";
    }

    {
        std::cout << "6. GmodNode: Mappability\n";
        std::cout << "--------------------------\n";

        const auto& vis = VIS::instance();
        const auto& gmod = vis.gmod(vis.latest());

        std::vector<std::string> testCodes = {
            "VE",      // Root - not mappable
            "411",     // Mappable
            "411.1",   // Not mappable
            "C101",    // Mappable
            "C101.31", // Mappable
            "CS1"      // Product selection - not mappable
        };

        std::cout << std::boolalpha;
        for (const auto& code : testCodes)
        {
            auto nodeOpt = gmod.node(code);
            if (nodeOpt.has_value())
            {
                std::cout << "  " << std::setw(7) << std::left << code;
                std::cout << ": " << (*nodeOpt)->isMappable() << "\n";
            }
        }

        std::cout << "\n";
    }

    {
        std::cout << "7. Gmod: Iterating all nodes\n";
        std::cout << "------------------------------\n";

        const auto& vis = VIS::instance();
        const auto& gmod = vis.gmod(vis.latest());

        std::cout << "First 10 nodes:\n";
        size_t count = 0;
        for (const auto& [code, node] : gmod)
        {
            if (count++ >= 10)
            {
                break;
            }
            std::cout << "  " << std::setw(8) << std::left << code;
            std::cout << " - " << node.metadata().name() << "\n";
        }

        std::cout << "\n";
    }

    {
        std::cout << "8. GmodNode: Finding specific node types\n";
        std::cout << "-------------------------------------------\n";

        const auto& vis = VIS::instance();
        const auto& gmod = vis.gmod(vis.latest());

        // Find PRODUCT TYPE nodes
        std::cout << "Sample PRODUCT TYPE nodes:\n";
        size_t count = 0;
        for (const auto& [code, node] : gmod)
        {
            if (node.metadata().category() == "PRODUCT" && node.metadata().type() == "TYPE")
            {
                if (count++ >= 5)
                {
                    break;
                }
                std::cout << "  " << std::setw(10) << std::left << code;
                std::cout << " - " << node.metadata().name() << "\n";
            }
        }

        // Find ASSET FUNCTION LEAF nodes
        std::cout << "\nSample ASSET FUNCTION LEAF nodes:\n";
        count = 0;
        for (const auto& [code, node] : gmod)
        {
            if (node.isLeafNode() && node.metadata().category() == "ASSET FUNCTION")
            {
                if (count++ >= 5)
                {
                    break;
                }
                std::cout << "  " << std::setw(10) << std::left << code;
                std::cout << " - " << node.metadata().name() << "\n";
            }
        }

        std::cout << "\n";
    }

    {
        std::cout << "9. GmodNode: Tree navigation example\n";
        std::cout << "----------------------------------------\n";

        const auto& vis = VIS::instance();
        const auto& gmod = vis.gmod(vis.latest());

        // Navigate from a node to root
        const auto& node = gmod["C101.31"];
        std::cout << "Path from " << node.code() << " to root:\n";

        const GmodNode* current = &node;
        int depth = 0;
        std::unordered_set<std::string_view> visited;
        while (current != nullptr)
        {
            if (!visited.insert(current->code()).second)
            {
                std::cout << std::string(depth * 2, ' ') << "- [cycle detected at " << current->code() << "]\n";
                break;
            }

            std::cout << std::string(depth * 2, ' ') << "- " << current->code();
            std::cout << " (" << current->metadata().name() << ")\n";

            current = current->parents().empty() ? nullptr : current->parents()[0];
            ++depth;
        }

        std::cout << "\n";
    }

    {
        std::cout << "10. GmodNode: Node properties comparison\n";
        std::cout << "-------------------------------------------\n";

        const auto& vis = VIS::instance();
        const auto& gmod = vis.gmod(vis.latest());

        std::vector<std::string> codes = { "411.1", "C101", "CS1", "F201" };

        std::cout << std::left;
        std::cout << std::setw(12) << "Code";
        std::cout << std::setw(20) << "Category";
        std::cout << std::setw(15) << "Type";
        std::cout << std::setw(12) << "Mappable";
        std::cout << std::setw(12) << "IsLeaf";
        std::cout << "\n";
        std::cout << std::string(71, '-') << "\n";

        std::cout << std::boolalpha;
        for (const auto& code : codes)
        {
            auto nodeOpt = gmod.node(code);
            if (nodeOpt.has_value())
            {
                const auto& node = **nodeOpt;
                std::cout << std::setw(12) << code;
                std::cout << std::setw(20) << node.metadata().category();
                std::cout << std::setw(15) << node.metadata().type();
                std::cout << std::setw(12) << node.isMappable();
                std::cout << std::setw(12) << node.isLeafNode();
                std::cout << "\n";
            }
        }

        std::cout << "\n";
    }

    {
        std::cout << "11. Gmod: Working with different versions\n";
        std::cout << "--------------------------------------------\n";

        const auto& vis = VIS::instance();

        // Compare node counts across versions
        std::cout << "Node counts across VIS versions:\n";
        for (const auto& version : vis.versions())
        {
            const auto& gmod = vis.gmod(version);
            size_t count = 0;
            for ([[maybe_unused]] const auto& [code, node] : gmod)
            {
                count++;
            }

            std::cout << "  " << std::setw(5) << std::left << VisVersions::toString(version);
            std::cout << ": " << std::setw(4) << std::right << count << " nodes\n";
        }

        std::cout << "\n";
    }

    return 0;
}

File: bindings/rust/vista-sdk/examples/core/gmod.rs
use vista_sdk::core::vis::Vis;

fn main() {
    println!("=== vista-sdk Gmod Sample ===\n");

    {
        println!("1. Gmod: Accessing Gmod for a VIS version");
        println!("--------------------------------------------");

        let vis = Vis::instance();
        let version = vis.latest();
        let gmod = vis
            .gmod(version)
            .expect("latest() should always be a valid VIS version");

        println!("Gmod for version   : {}", gmod.version().as_str());

        let node_count = gmod.iter().count();

        println!("Total nodes in Gmod: {}", node_count);
        println!("Root node code     : {}", gmod.root_node().code());
        println!();
    }

    {
        println!("2. GmodNode: Accessing nodes by code");
        println!("---------------------------------------");

        let vis = Vis::instance();
        let gmod = vis
            .gmod(vis.latest())
            .expect("latest() should always be a valid VIS version");

        // Access node using get_node
        let node1 = gmod.get_node("411.1").expect("411.1 should exist");
        println!("Node code    : {}", node1.code());
        println!("Node name    : {}", node1.metadata().name());
        println!("Node category: {}", node1.metadata().category());
        println!("Node type    : {}", node1.metadata().r#type());

        // Safe access using get_node
        if let Ok(node2) = gmod.get_node("C101.31") {
            println!("\nFound node: {}", node2.code());
            println!("Name      : {}", node2.metadata().name());
        }

        // Try invalid code
        let invalid = gmod.get_node("INVALID");
        println!("\nInvalid code lookup: {}", invalid.is_ok());

        println!();
    }

    {
        println!("3. GmodNode: Navigating the tree hierarchy");
        println!("---------------------------------------------");

        let vis = Vis::instance();
        let gmod = vis
            .gmod(vis.latest())
            .expect("latest() should always be a valid VIS version");

        let node = gmod.get_node("411.1").expect("411.1 should exist");

        // Parents (can have multiple in Gmod graph)
        if node.parent_count() > 0 {
            println!("Parents of {} ({}):", node.code(), node.parent_count());
            for parent in node.parents() {
                println!("  - {}", parent.code());
            }
        }

        // Children
        println!("\nChildren of {} ({}):", node.code(), node.child_count());
        let max_display = 5;
        for (count, child) in node.children().enumerate() {
            if count >= max_display {
                println!("  ... and {} more", node.child_count() - max_display);
                break;
            }
            println!("  - {} ({})", child.code(), child.metadata().name());
        }

        // Root check
        println!("\nIs root  ? {}", node.is_root());
        println!("Root node: {}", gmod.root_node().code());
        println!("Is root  ? {}", gmod.root_node().is_root());

        println!();
    }

    {
        println!("4. GmodNode: Node metadata");
        println!("----------------------------");

        let vis = Vis::instance();
        let gmod = vis
            .gmod(vis.latest())
            .expect("latest() should always be a valid VIS version");

        let node = gmod.get_node("S222.11").expect("S222.11 should exist");
        let metadata = node.metadata();

        println!("Node                : {}", node.code());
        println!("Category            : {}", metadata.category());
        println!("Type                : {}", metadata.r#type());
        println!("Name                : {}", metadata.name());

        if let Some(definition) = metadata.definition() {
            println!("Definition      : {}", definition);
        }

        if let Some(install_substructure) = metadata.install_substructure() {
            println!("Install substructure: {}", install_substructure);
        }

        println!();
    }

    {
        println!("5. GmodNode: Product types and selections");
        println!("--------------------------------------------");

        let vis = Vis::instance();
        let gmod = vis
            .gmod(vis.latest())
            .expect("latest() should always be a valid VIS version");

        // Product type example
        let node1 = gmod.get_node("411.3").expect("411.3 should exist");
        println!("Node             : {}", node1.code());
        println!("Has product type : {}", node1.product_type().is_some());
        if let Some(prod_type) = node1.product_type() {
            println!("Product type code: {}", prod_type.code());
            println!("Product type name: {}", prod_type.metadata().name());
        }

        // Product selection example
        let node2 = gmod.get_node("411.2").expect("411.2 should exist");
        println!("\nNode: {}", node2.code());
        println!(
            "Has product selection : {}",
            node2.product_selection().is_some()
        );
        if let Some(prod_sel) = node2.product_selection() {
            println!("Product selection code: {}", prod_sel.code());
            println!("Product selection name: {}", prod_sel.metadata().name());
        }

        // Check if node is product selection
        if let Ok(cs_node) = gmod.get_node("CS1") {
            println!("\nNode: {}", cs_node.code());
            println!("Is product selection: {}", cs_node.is_product_selection());
        }

        println!();
    }

    {
        println!("6. GmodNode: Mappability");
        println!("--------------------------");

        let vis = Vis::instance();
        let gmod = vis
            .gmod(vis.latest())
            .expect("latest() should always be a valid VIS version");

        let test_codes = [
            "VE",      // Root - not mappable
            "411",     // Mappable
            "411.1",   // Not mappable
            "C101",    // Mappable
            "C101.31", // Mappable
            "CS1",     // Product selection - not mappable
        ];

        for code in test_codes {
            if let Ok(node) = gmod.get_node(code) {
                println!("  {:<7}: {}", code, node.is_mappable());
            }
        }

        println!();
    }

    {
        println!("7. Gmod: Iterating all nodes");
        println!("------------------------------");

        let vis = Vis::instance();
        let gmod = vis
            .gmod(vis.latest())
            .expect("latest() should always be a valid VIS version");

        println!("First 10 nodes:");
        for (count, node) in gmod.iter().enumerate() {
            if count >= 10 {
                break;
            }
            println!("  {:<8} - {}", node.code(), node.metadata().name());
        }

        println!();
    }

    {
        println!("8. GmodNode: Finding specific node types");
        println!("-------------------------------------------");

        let vis = Vis::instance();
        let gmod = vis
            .gmod(vis.latest())
            .expect("latest() should always be a valid VIS version");

        // Find PRODUCT TYPE nodes
        println!("Sample PRODUCT TYPE nodes:");
        let mut count = 0;
        for node in gmod.iter() {
            if node.metadata().category() == "PRODUCT" && node.metadata().r#type() == "TYPE" {
                if count >= 5 {
                    break;
                }
                count += 1;
                println!("  {:<10} - {}", node.code(), node.metadata().name());
            }
        }

        // Find ASSET FUNCTION LEAF nodes
        println!("\nSample ASSET FUNCTION LEAF nodes:");
        let mut count = 0;
        for node in gmod.iter() {
            if node.is_leaf_node() && node.metadata().category() == "ASSET FUNCTION" {
                if count >= 5 {
                    break;
                }
                count += 1;
                println!("  {:<10} - {}", node.code(), node.metadata().name());
            }
        }

        println!();
    }

    {
        println!("9. GmodNode: Tree navigation example");
        println!("----------------------------------------");

        let vis = Vis::instance();
        let gmod = vis
            .gmod(vis.latest())
            .expect("latest() should always be a valid VIS version");

        // Navigate from a node to root
        let node = gmod.get_node("C101.31").expect("C101.31 should exist");
        println!("Path from {} to root:", node.code());

        let mut current = Some(node);
        let mut depth = 0;
        let mut visited = std::collections::HashSet::new();
        while let Some(n) = current {
            if !visited.insert(n.code()) {
                println!(
                    "{}- [cycle detected at {}]",
                    " ".repeat(depth * 2),
                    n.code()
                );
                break;
            }

            println!(
                "{}- {} ({})",
                " ".repeat(depth * 2),
                n.code(),
                n.metadata().name()
            );

            // Navigate to first parent (if any)
            current = n.parent_at(0);
            depth += 1;
        }

        println!();
    }

    {
        println!("10. GmodNode: Node properties comparison");
        println!("-------------------------------------------");

        let vis = Vis::instance();
        let gmod = vis
            .gmod(vis.latest())
            .expect("latest() should always be a valid VIS version");

        let codes = ["411.1", "C101", "CS1", "F201"];

        println!(
            "{:<12}{:<20}{:<15}{:<12}{:<12}",
            "Code", "Category", "Type", "Mappable", "IsLeaf"
        );
        println!("{}", "-".repeat(71));

        for code in codes {
            if let Ok(node) = gmod.get_node(code) {
                println!(
                    "{:<12}{:<20}{:<15}{:<12}{:<12}",
                    code,
                    node.metadata().category(),
                    node.metadata().r#type(),
                    node.is_mappable(),
                    node.is_leaf_node()
                );
            }
        }

        println!();
    }

    {
        println!("11. Gmod: Working with different versions");
        println!("--------------------------------------------");

        let vis = Vis::instance();

        // Compare node counts across versions
        println!("Node counts across VIS versions:");
        for version in vis.versions() {
            let gmod = vis
                .gmod(version)
                .expect("versions() only returns valid VIS versions");
            let count = gmod.iter().count();

            println!("  {:<5}: {:>4} nodes", version.as_str(), count);
        }

        println!();
    }
}
